clippy_lints/src/methods/mod.rs

```diff
@@ -4339,6 +4340,34 @@ declare_clippy_lint! {
     "using `NonZero::new_unchecked()` in a `const` context"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for string slices immediantly followed by `as_bytes`.
+    ///
+    /// ### Why is this bad?
+    /// It involves doing an unnecessary UTF-8 alignment check which is less efficient, and can cause a panic.
+    ///
+    /// ### Known problems
+    /// In some cases, the UTF-8 validation and potential panic from string slicing may be required for
+    /// the code's correctness. If you need to ensure the slice boundaries fall on valid UTF-8 character
+    /// boundaries, the original form (`s[1..5].as_bytes()`) should be preferred.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let s = "Lorem ipsum";
+    /// s[1..5].as_bytes();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let s = "Lorem ipsum";
+    /// &s.as_bytes()[1..5];
+    /// ```
+     #[clippy::version = "1.86.0"]
+     pub SLICED_STRING_AS_BYTES,
+     pedantic,
+     "slicing a string and immediately calling as_bytes is less efficient and can lead to panics"
```

Comment:
Nit: shouldn't we use something like "lead to panics when slicing inside a multi-byte char", or "outside of char boundary"? And if it is not in "perf" (and "pedantic" is fine with me here), you may drop the "less efficient".

What do others think?